{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Naive exact matching plus reverse complement"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The goal is to implement the naive exact matching algorithm that is strand-aware. Meaning, it will look for occurrences of the target sequence in the genome, along with the reverse complement of the sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFastq(filename):\n",
    "    '''Parses read and quality strings from FASTQ file format.'''\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()  # skip name line\n",
    "            seq = fh.readline().rstrip()  # read base sequence\n",
    "            fh.readline()  # skip placeholder line\n",
    "            qual = fh.readline().rstrip() # base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverseComplement(sequence):\n",
    "    '''Creates reverse complement sequence from input sequence'''\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N'}\n",
    "    reversed_seq = ''\n",
    "    for base in sequence:\n",
    "        reversed_seq = complement[base] + reversed_seq\n",
    "    return reversed_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readGenome(filename):\n",
    "    '''Parses DNA sequence from FASTA file format.'''\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse(s):\n",
    "  str = \"\"\n",
    "  for i in s:\n",
    "    str = i + str\n",
    "  return str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def naive(pattern, text):\n",
    "    occurrences = []\n",
    "    for i in range(len(text) - len(pattern) + 1):  # loop over alignments\n",
    "        match = True\n",
    "        for j in range(len(pattern)):  # loop over characters\n",
    "            if text[i+j] != pattern[j]:  # compare characters\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def naive_2mm(pattern, text):\n",
    "    occurrences = []\n",
    "    for i in range(len(text) - len(pattern) + 1):  # loop over alignments\n",
    "        match = True\n",
    "        mismatch = 0\n",
    "        for j in range(len(pattern)):  # loop over characters\n",
    "            if text[i+j] != pattern[j]:  # compare characters\n",
    "                mismatch += 1\n",
    "            if mismatch > 2:\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def naive_with_rc(pattern, text):\n",
    "    '''Execute exact matching algorithm on given pattern and its reverse complement.'''\n",
    "    occurrences = []\n",
    "    rev_comp = reverseComplement(p)\n",
    "    check_equal = reverse(rev_comp)\n",
    "    \n",
    "    for i in range(len(text) - len(pattern) + 1):  # loop over alignments\n",
    "        match = True  # initialize to true\n",
    "        for j in range(len(pattern)):  # loop over characters\n",
    "            if pattern != check_equal:\n",
    "                if text[i+j] != pattern[j] and text[i+j] != rev_comp[j]:  # compare characters\n",
    "                    match = False\n",
    "                    break\n",
    "            elif text[i+j] != pattern[j]:\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "def phred33ToQ(qual):\n",
    "    return ord(qual) - 33"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def createHist(qualities):\n",
    "    hist = [0] * 50  # highest is 41, 50 for padding\n",
    "    for qual in qualities:\n",
    "        for phred in qual:\n",
    "            q = phred33ToQ(phred)\n",
    "            hist[q] += 1\n",
    "    return hist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[10, 23]\n"
     ]
    }
   ],
   "source": [
    "# test scenario 1\n",
    "p = 'CCC'\n",
    "ten_as = 'AAAAAAAAAA'\n",
    "t = ten_as + 'CCC' + ten_as + 'GGG' + ten_as\n",
    "occurrences = naive_with_rc(p, t)\n",
    "print(occurrences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[10, 24]\n"
     ]
    }
   ],
   "source": [
    "# test scenario 2\n",
    "p = 'CGCG'\n",
    "t = ten_as + 'CGCG' + ten_as + 'CGCG' + ten_as\n",
    "occurrences = naive_with_rc(p, t)\n",
    "print(occurrences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test scenario 3\n",
    "# Phi-X genome\n",
    "# !wget http://d396qusza40orc.cloudfront.net/ads1/data/phix.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "phix_genome = readGenome('phix.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "occurrences = naive_with_rc('ATTA', phix_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "offset of leftmost occurrence: 7\n",
      "# occurrences: 285\n"
     ]
    }
   ],
   "source": [
    "print('offset of leftmost occurrence: %d' % min(occurrences))\n",
    "print('# occurrences: %d' % len(occurrences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download and parse lambda virus\n",
    "# !wget https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/lambda_virus.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "lambda_genome = readGenome('lambda_virus.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "306\n"
     ]
    }
   ],
   "source": [
    "# question 1\n",
    "p1 = 'AGGT'\n",
    "q1_occurrences = naive(p1, lambda_genome)\n",
    "q1_occurrences2 = naive(reverseComplement(p1), lambda_genome)\n",
    "print(len(q1_occurrences) + len(q1_occurrences2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "195\n"
     ]
    }
   ],
   "source": [
    "# question 2\n",
    "p2 = 'TTAA'\n",
    "q2_occurrences = naive(p2, lambda_genome)\n",
    "print(len(q2_occurrences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "offset of leftmost occurrence: 27733\n",
      "offset of leftmost occurrence: 26028\n"
     ]
    }
   ],
   "source": [
    "# question 3\n",
    "p3 = 'ACTAAGT'\n",
    "q3_occurrences = naive(p3, lambda_genome)\n",
    "print(f'offset of leftmost occurrence: {min(q3_occurrences)}')\n",
    "q3_occurrences2 = naive(reverseComplement(p3), lambda_genome)\n",
    "print(f'offset of leftmost occurrence: {min(q3_occurrences2)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "offset of leftmost occurrence: 18005\n",
      "offset of leftmost occurrence: 450\n"
     ]
    }
   ],
   "source": [
    "# question 4\n",
    "p4 = 'AGTCGA'\n",
    "q4_occurrences = naive(p4, lambda_genome)\n",
    "print(f'offset of leftmost occurrence: {min(q4_occurrences)}')\n",
    "q4_occurrences2 = naive(reverseComplement(p4), lambda_genome)\n",
    "print(f'offset of leftmost occurrence: {min(q4_occurrences2)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "191\n"
     ]
    }
   ],
   "source": [
    "# question 5\n",
    "p5 = 'TTCAAGCC'\n",
    "q5_occurrences = naive_2mm(p5, lambda_genome)\n",
    "print(len(q5_occurrences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "offset of leftmost occurrence: 49\n"
     ]
    }
   ],
   "source": [
    "# question 6\n",
    "p6 = 'AGGAGGTT'\n",
    "q6_occurrences = naive_2mm(p6, lambda_genome)\n",
    "print(f'offset of leftmost occurrence: {min(q6_occurrences)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 39, 40, 39, 36, 39, 38, 39, 37, 39, 39, 28, 2, 28, 29, 29, 37, 37, 37, 37, 33, 40, 40, 39, 38, 36, 39, 36, 36, 38, 38, 36, 39, 29, 39, 38, 30, 25, 35, 30, 35, 33, 35, 37, 37, 33]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXAAAAD4CAYAAAD1jb0+AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAPq0lEQVR4nO3dbYxcZ3nG8f9VOxDeWidkZbl21E0hAkVIOGjrBgUhaqAyCSJBiioiRP0hlalE1NCiUkM/FKRWChIQqFRFMjjErWgKDaGJEkqbmiCE1JquwRgnhiaEUGw58SIIhH4AnNz9MMewrHe9szszu35m/z9pNOc858yc+8njXDp7XuakqpAktefXVrsASdLyGOCS1CgDXJIaZYBLUqMMcElq1PqV3NhFF11Uk5OTK7lJSWrewYMHv19VE3PbVzTAJycnmZ6eXslNSlLzknx3vnYPoUhSowxwSWqUAS5JjTLAJalRBrgkNcoAl6RG9R3gSdYl+VqSe7v5S5IcSPJIkk8ledboypQkzbWUPfCbgKOz5j8A3FJVLwZ+CNwwzMIkSWfXV4An2QJcDXy8mw+wHbizW2UfcO0I6pMkLaDfPfCPAO8GnunmXwg8WVWnuvljwOb5PphkV5LpJNMzMzOD1Co1YXL3fUzuvm+1yxi52f1cK30+1ywa4EneCJysqoPL2UBV7amqqaqampg441Z+SdIy9fNbKFcCb0pyFXA+8OvAR4ENSdZ3e+FbgOOjK1OSNNeie+BV9Z6q2lJVk8BbgC9U1VuBB4DrutV2AnePrEpJ0hkGuQ78L4A/S/IIvWPie4dTkiSpH0v6Odmq+iLwxW76UWDb8Eua3yhOkDx289W/+N5zYXolnAv9XOn+r3R/Zlusfe6yftrn9u1cs1Cf51tnsfX63dZ839NPHa3zTkxJapQBLkmNMsAlqVEGuCQ1akWfiSnpTON8B+MgJxKHeeJ2XE9ougcuSY0ywCWpUQa4JDXKAJekRnkSUzpHrcTJzUHuMtUvrdZ/L/fAJalRBrgkNcoAl6RGGeCS1ChPYkorZK2fDFzopOxST9aO852rS+UeuCQ1qp+HGp+f5CtJvp7kwSTv79pvT/KdJIe619aRVytJ+oV+DqH8FNheVT9Jch7w5ST/2i3786q6c3TlSZIWsmiAV1UBP+lmz+teNcqiJEmL6+sYeJJ1SQ4BJ4H7q+pAt+hvkhxOckuSZ4+qSEnSmfoK8Kp6uqq2AluAbUleBrwHeCnwO8CF9J5Sf4Yku5JMJ5memZkZTtWSNGSTu++b9wqXpbavpCVdhVJVTwIPADuq6kT1/BT4BAs8ob6q9lTVVFVNTUxMDFywJKmnn6tQJpJs6KafA7we+GaSTV1bgGuBI6MrU5I0Vz9XoWwC9iVZRy/wP11V9yb5QpIJIMAh4I9HV6Ykaa5+rkI5DFw+T/v2kVQkSeqLd2JKUqMMcElqlAEuSY0ywCWpUQa4JDXK3wOX1Dd/03xp/Z99p+Yo/pu5By5JjTLAJalRBrgkNcoAl6RGGeCS1CgDXJIaZYBLUqMMcElqlAEuSY0ywCWpUQa4JDWqn2dinp/kK0m+nuTBJO/v2i9JciDJI0k+leRZoy9XknRaP3vgPwW2V9XLga3AjiRXAB8AbqmqFwM/BG4YWZWSpDMsGuDV85Nu9rzuVcB24M6ufR+9J9NLklZIX8fAk6xLcgg4CdwPfBt4sqpOdascAzYv8NldSaaTTM/MzAyhZElavsnd9/3Kz7wud51zQV8BXlVPV9VWYAuwDXhpvxuoqj1VNVVVUxMTE8urUpJ0hiVdhVJVTwIPAK8ENiQ5/UCILcDx4ZYmSTqbfq5CmUiyoZt+DvB64Ci9IL+uW20ncPeIapQkzaOfR6ptAvYlWUcv8D9dVfcmeQj4pyR/DXwN2DvCOiVJcywa4FV1GLh8nvZH6R0PlyStAh9qLK0Bsx/GO64PJh71A4TPRd5KL0mNMsAlqVEGuCQ1ygCXpEZ5ElNq2NlO3I3ryUr9knvgktQoA1ySGmWAS1KjDHBJapQBLkmNMsAlqVEGuCQ1ygCXpEYZ4JLUKANckhplgEtSo/p5JubFSR5I8lCSB5Pc1LW/L8nxJIe611WjL1eSdFo/P2Z1CnhXVX01yQuAg0nu75bdUlUfHF15kqSF9PNMzBPAiW76qSRHgc2jLkySdHZLOgaeZJLeA44PdE03Jjmc5LYkFyzwmV1JppNMz8zMDFatJA3R5O77fuUneVvTd4AneT7wGeCdVfVj4FbgRcBWenvoH5rvc1W1p6qmqmpqYmJi8IolSUCfAZ7kPHrh/cmqugugqp6oqqer6hngY8C20ZUpSZqrn6tQAuwFjlbVh2e1b5q12puBI8MvT5K0kH6uQrkSeBvwjSSHurb3Atcn2QoU8Bjw9hHUJ0laQD9XoXwZyDyLPjf8ciRJ/fJOTElqlAEuSY0ywCWpUQa4JDXKAJekRhngktQoA1ySGmWAS1KjDHBJapQBLkmNMsAlqVEGuCQ1ygCXpEYZ4JLUKANckhplgEtSowxwSWpUP8/EvDjJA0keSvJgkpu69guT3J/k4e79gtGXK0k6rZ898FPAu6rqMuAK4B1JLgN2A/ur6lJgfzcvSVohiwZ4VZ2oqq92008BR4HNwDXAvm61fcC1I6pRkjSPJR0DTzIJXA4cADZW1Ylu0ePAxgU+syvJdJLpmZmZQWqVJM3Sd4AneT7wGeCdVfXj2cuqqoCa73NVtaeqpqpqamJiYqBiJUm/1FeAJzmPXnh/sqru6pqfSLKpW74JODmaEiVJ8+nnKpQAe4GjVfXhWYvuAXZ20zuBu4dfniRpIev7WOdK4G3AN5Ic6treC9wMfDrJDcB3gT8YSYWSpHktGuBV9WUgCyx+7XDLkST1yzsxJalRBrgkNcoAl6RGGeCS1CgDXJIaZYBLUqMMcElqlAEuSY0ywCWpUQa4JDXKAJekRhngktQoA1ySGmWAS1KjDHBJapQBLkmNMsAlqVH9PBPztiQnkxyZ1fa+JMeTHOpeV422TEnSXP3sgd8O7Jin/Zaq2tq9PjfcsiRJi1k0wKvqS8APVqAWSdISDHIM/MYkh7tDLBcstFKSXUmmk0zPzMwMsDlJ0mzLDfBbgRcBW4ETwIcWWrGq9lTVVFVNTUxMLHNzkqS5lhXgVfVEVT1dVc8AHwO2DbcsSdJilhXgSTbNmn0zcGShdSVJo7F+sRWS3AG8BrgoyTHgr4DXJNkKFPAY8PbRlShJms+iAV5V18/TvHcEtUiSlsA7MSWpUQa4JDXKAJekRhngktQoA1ySGmWAS1KjDHBJapQBLkmNMsAlqVEGuCQ1ygCXpEYZ4JLUKANckhplgEtSowxwSWqUAS5JjTLAJalRiwZ4ktuSnExyZFbbhUnuT/Jw937BaMuUJM3Vzx747cCOOW27gf1VdSmwv5uXJK2gRQO8qr4E/GBO8zXAvm56H3DtcMuSJC1mucfAN1bViW76cWDjQism2ZVkOsn0zMzMMjcnSZpr4JOYVVVAnWX5nqqaqqqpiYmJQTcnSeosN8CfSLIJoHs/ObySJEn9WG6A3wPs7KZ3AncPpxxJUr/6uYzwDuA/gZckOZbkBuBm4PVJHgZe181LklbQ+sVWqKrrF1j02iHXIklaAu/ElKRGGeCS1CgDXJIaZYBLUqMMcElqlAEuSY0ywCWpUQa4JDXKAJekRhngktQoA1ySGmWAS1KjDHBJapQBLkmNMsAlqVEGuCQ1ygCXpEYt+kSes0nyGPAU8DRwqqqmhlGUJGlxAwV45/eq6vtD+B5J0hJ4CEWSGjVogBfw70kOJtk13wpJdiWZTjI9MzMz4OYkSacNGuCvqqpXAG8A3pHk1XNXqKo9VTVVVVMTExMDbk6SdNpAAV5Vx7v3k8BngW3DKEqStLhlB3iS5yV5welp4PeBI8MqTJJ0doNchbIR+GyS09/zj1X1+aFUJUla1LIDvKoeBV4+xFokSUvgZYSS1CgDXJIaZYBLUqMMcElqlAEuSY0ywCWpUQa4JDXKAJekRhngktSoYTzQQdI5YHL3fb+Yfuzmq1exEq0U98AlqVEGuCQ1ygCXpEYZ4JLUKANckhplgEtSowxwSWrUQAGeZEeSbyV5JMnuYRUlSVrcIA81Xgf8HfAG4DLg+iSXDaswSdLZDbIHvg14pKoeraqfAf8EXDOcsiRJi0lVLe+DyXXAjqr6o27+bcDvVtWNc9bbBezqZl8CfGv55XIR8P0BPt8i+7w22Oe1Ybl9/q2qmpjbOPLfQqmqPcCeYXxXkumqmhrGd7XCPq8N9nltGHafBzmEchy4eNb8lq5NkrQCBgnw/wYuTXJJkmcBbwHuGU5ZkqTFLPsQSlWdSnIj8G/AOuC2qnpwaJXNbyiHYhpjn9cG+7w2DLXPyz6JKUlaXd6JKUmNMsAlqVHNBPhauG0/ycVJHkjyUJIHk9zUtV+Y5P4kD3fvF6x2rcOWZF2SryW5t5u/JMmBbrw/1Z0oHxtJNiS5M8k3kxxN8spxH+ckf9r9uz6S5I4k54/bOCe5LcnJJEdmtc07run5267vh5O8YqnbayLA19Bt+6eAd1XVZcAVwDu6fu4G9lfVpcD+bn7c3AQcnTX/AeCWqnox8EPghlWpanQ+Cny+ql4KvJxe38d2nJNsBv4EmKqql9G78OEtjN843w7smNO20Li+Abi0e+0Cbl3qxpoIcNbIbftVdaKqvtpNP0Xvf+rN9Pq6r1ttH3DtqhQ4Ikm2AFcDH+/mA2wH7uxWGas+J/kN4NXAXoCq+llVPcmYjzO9q96ek2Q98FzgBGM2zlX1JeAHc5oXGtdrgL+vnv8CNiTZtJTttRLgm4HvzZo/1rWNrSSTwOXAAWBjVZ3oFj0ObFytukbkI8C7gWe6+RcCT1bVqW5+3Mb7EmAG+ER32OjjSZ7HGI9zVR0HPgj8L73g/hFwkPEe59MWGteBc62VAF9Tkjwf+Azwzqr68exl1bvuc2yu/UzyRuBkVR1c7VpW0HrgFcCtVXU58H/MOVwyhuN8Ab09zkuA3wSex5mHGsbesMe1lQBfM7ftJzmPXnh/sqru6pqfOP2nVfd+crXqG4ErgTcleYzeobHt9I4Pb+j+1IbxG+9jwLGqOtDN30kv0Md5nF8HfKeqZqrq58Bd9MZ+nMf5tIXGdeBcayXA18Rt+92x373A0ar68KxF9wA7u+mdwN0rXduoVNV7qmpLVU3SG9cvVNVbgQeA67rVxq3PjwPfS/KSrum1wEOM8TjTO3RyRZLndv/OT/d5bMd5loXG9R7gD7urUa4AfjTrUEt/qqqJF3AV8D/At4G/XO16RtTHV9H78+owcKh7XUXvmPB+4GHgP4ALV7vWEfX/NcC93fRvA18BHgH+GXj2atc35L5uBaa7sf4X4IJxH2fg/cA3gSPAPwDPHrdxBu6gd4z/5/T+0rphoXEFQu/qum8D36B3hc6Stuet9JLUqFYOoUiS5jDAJalRBrgkNcoAl6RGGeCS1CgDXJIaZYBLUqP+HxIRe1W8IV4tAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "66\n"
     ]
    }
   ],
   "source": [
    "# question 7\n",
    "#!wget https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR037900_1.first1000.fastq\n",
    "\n",
    "def createHist(qualities):\n",
    "    sequences, qualities = readFastq('ERR037900_1.first1000.fastq')\n",
    "    phredscore = []\n",
    "    qualities = qualities[-5]\n",
    "    for phred in qualities:\n",
    "        q = ord(phred)-33\n",
    "        phredscore.append(q)\n",
    "    return phredscore\n",
    "h = createHist(qualities)\n",
    "print (h)\n",
    "import matplotlib.pyplot as plt\n",
    "plt.bar(range(len(h)), h)\n",
    "plt.show()\n",
    "min_h = min(h)\n",
    "index_min_h = h.index(min_h)\n",
    "print(index_min_h)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:genomics]",
   "language": "python",
   "name": "conda-env-genomics-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
